### PR TITLE
Uhm 5842 1

### PR DESCRIPTION
**Pull request by mseaton**
Adds support for 2 new job types (sql-execution and iterating-job), which allow more flexibility of operations and flow, and more convenient and maintainable configuration where the same jobs need to be executed multiple times with small variations (eg. the source datasource).  Also adds in support for startupJobs, which are jobs that are always run at the time of application startup and before any scheduled job executions.  Finally adds in support within the iterating-job and job-pipeline job (which both execute multiple jobs) to determine whether these jobs run in series or in parallel, the number of concurrent threads to use in parallel, and whether a given job should be re-tried on failure.

### CODE REVIEW

File: src/main/java/org/pih/petl/api/ExecutionContext.java [hunk 1]
```diff
@@ -25,6 +32,29 @@ public ExecutionContext(JobExecution jobExecution, PetlJobConfig jobConfig, Appl
...
+            if (replacementVariables != null) {
+                replacements.putAll(replacementVariables);
+            }
+            configuration = StrSubstitutor.replace(configuration, replacements);
+            JobConfiguration jobConfiguration = new JobConfiguration(PetlUtil.readJsonFromString(configuration));
+            jobConfiguration.setVariables(replacements);
+            nestedJobConfig.setConfiguration(jobConfiguration);
+            return nestedJobConfig;
+        }
+    }
```
Comment (mseaton) [inline src/main/java/org/pih/petl/api/ExecutionContext.java:56]: This is a convenience method within the Execution Context to get a nested Job, in order that the job-pipeline and iterating-job implementations can both leverage this to run their nested jobs and take advantage of the same features (like variable replacement).

File: src/main/java/org/pih/petl/api/JobExecutionResult.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    public Throwable getException() {
+        return exception;
+    }
+
+    public void setException(Throwable exception) {
+        this.exception = exception;
+    }
+}
```
Comment (mseaton) [inline src/main/java/org/pih/petl/api/JobExecutionResult.java:29]: This is what the JobExecutionTask returns (see below)

File: src/main/java/org/pih/petl/api/JobExecutionTask.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                }
+                else {
+                    context.setStatus(status.toString());
+                    log.error(t);
+                }
+            }
+        }
+        return result;
+    }
+}
```
Comment (mseaton) [inline src/main/java/org/pih/petl/api/JobExecutionTask.java:70]: This is what I came up with to represent a Thread that executes a job, and runs at a repeated interval in case of errors.  See:  
* https://howtodoinjava.com/java/multi-threading/executorservice-invokeall/
* https://www.baeldung.com/java-executor-service-tutorial
* https://howtodoinjava.com/java/multi-threading/java-thread-pool-executor-example/
* https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#invokeAll-java.util.Collection-

File: src/main/java/org/pih/petl/api/JobExecutor.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+        for (Future<JobExecutionResult> resultFutures : results) {
+            JobExecutionResult result = resultFutures.get();
+            if (!result.isSuccessful()) {
+                throw new RuntimeException("Error in one of the jobs", result.getException());
+            }
+        }
+    }
+
+}
```
Comment (mseaton) [inline src/main/java/org/pih/petl/api/JobExecutor.java:32]: See links above related to Executor.

File: src/main/java/org/pih/petl/job/config/ExecutionConfig.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.sql.Time;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Configures how execution threading, concurrency, and retries on error should be handled
+ * By default, an ExecutionConfig will specify no retries, and one concurrent job,
+ * but this can be adjusted to support executing jobs in parallel, and configuring the number of times a job
+ * execution should be attempted (i.e. to allow for retrying due to connectivity problems)
+ */
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/config/ExecutionConfig.java:11]: This is just a convenient way to read in this configuration and pass it around from jobs that support it (job-pipeline, iterating-job)

File: src/main/java/org/pih/petl/job/config/JobConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+import org.pih.petl.PetlUtil;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+/**
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/config/JobConfiguration.java:18]: This is just a formal Object to represent a job's configuration, which was previously just a JsonNode on the object, and I moved all of the various methods that previously existed on PetlJobConfig for getting configuration values from this onto this class.  See below.

File: src/main/java/org/pih/petl/job/config/PetlJobConfig.java [hunk 7]
```diff
@@ -160,11 +52,11 @@ public void setSchedule(Schedule schedule) {
         this.schedule = schedule;
     }
 
-    public JsonNode getConfiguration() {
+    public JobConfiguration getConfiguration() {
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/config/PetlJobConfig.java:55]: See comments above indicating how this was refactored.

File: src/main/java/org/pih/petl/job/type/IteratingJob.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    public void execute(final ExecutionContext context) throws Exception {
+        context.setStatus("Executing Iterating Job");
+        JobConfiguration config = context.getJobConfig().getConfiguration();
+
+        JsonNode jobTemplate = config.get("jobTemplate");
+        List<JsonNode> iterations = config.getList("iterations");
+        ExecutionConfig executionConfig = config.getObject(ExecutionConfig.class, "execution");
+        if (executionConfig == null) {
+            executionConfig = new ExecutionConfig(1, 0, 5, TimeUnit.MINUTES);
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/type/IteratingJob.java:41]: By default, we run with a maxConcurrent = 1, and no repeats.  I believe this will set up serial processing of jobs (though they will still run in separate threads).

File: src/main/java/org/pih/petl/job/type/RunMultipleJob.java [hunk 9]
```diff
@@ -28,22 +33,21 @@ public RunMultipleJob() {
...
-            PetlJob job = PetlJobFactory.instantiate(nestedJobConfig);
-            job.execute(nestedContext);
-            // TODO: Periodically update overall context status from nested status, using thread.  See SqlServerImportJob
-            context.setTotalLoaded(context.getTotalLoaded() + 1);
+        JobConfiguration config = context.getJobConfig().getConfiguration();
+
+        List<JsonNode> jobTemplates = config.getList("jobs");
+        ExecutionConfig executionConfig = config.getObject(ExecutionConfig.class, "execution");
+        if (executionConfig == null) {
+            executionConfig = new ExecutionConfig(1, 0, 5, TimeUnit.MINUTES);
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/type/RunMultipleJob.java:41]: See above.  This is backwards-incompatible with the previous version of this job type, but I don't think anyone has ever actually used this.  I'm bumping up the PETL version to 3.0.0 anyway to reflect this.

File: src/main/java/org/pih/petl/job/type/SqlServerImportJob.java [hunk 10]
```diff
@@ -65,13 +65,13 @@ public void execute(final ExecutionContext context) throws Exception {
         String sourceContextFileName = config.getString("extract", "context");
         if (sourceContextFileName != null) {
             ConfigFile sourceContextFile = appConfig.getConfigFile(sourceContextFileName);
-            sourceContextStatements = sourceContextFile.getContents();
+            sourceContextStatements = sourceContextFile.getContentsWithVariableReplacement(config.getVariables());
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/type/SqlServerImportJob.java:68]: Adding this in allows us to support variable replacement within the context sql file.

File: src/main/java/org/pih/petl/job/type/SqlServerImportJob.java [hunk 11]
```diff
@@ -65,13 +65,13 @@ public void execute(final ExecutionContext context) throws Exception {
...
         }
 
         // Get source query
         String sourceQueryFileName = config.getString("extract", "query");
         ConfigFile sourceQueryFile = appConfig.getConfigFile(sourceQueryFileName);
-        String sourceQuery = sourceContextStatements + sourceQueryFile.getContents();
+        String sourceQuery = sourceContextStatements + sourceQueryFile.getContentsWithVariableReplacement(config.getVariables());
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/type/SqlServerImportJob.java:74]: Adding this in allows us to support variable replacement within the source query sql file.

File: src/main/java/org/pih/petl/job/type/SqlServerImportJob.java [hunk 12]
```diff
@@ -84,17 +84,17 @@ public void execute(final ExecutionContext context) throws Exception {
         String targetTable = config.getString("load", "table");
 
         // Get target table schema
+        String targetSchema = null;
         String targetSchemaFilename = config.getString("load", "schema");
-        ConfigFile targetSchemaFile = appConfig.getConfigFile(targetSchemaFilename);
-        String targetSchema = targetSchemaFile.getContents();
-
-        // TODO: Add validation in
+        if (StringUtils.isNotEmpty(targetSchemaFilename)) {
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/type/SqlServerImportJob.java:89]: This just allows the schema to be optional.  It may be that we don't want the job to take responsibility over creating the schema, but may want that to happen explicitly via an earlier job.

File: src/main/java/org/pih/petl/job/type/SqlServerImportJob.java [hunk 13]
```diff
@@ -113,38 +113,40 @@ public void execute(final ExecutionContext context) throws Exception {
                 sourceConnection.setAutoCommit(false); // We intend to rollback changes to source after querying DB
                 targetConnection.setAutoCommit(true);  // We want to commit to target as we go, to query status
 
-                if (config.getBoolean(true,"dropAndRecreateTable")) {
-                    // drop existing target table  (we don't use "drop table if exists..." syntax for backwards compatibility with earlier versions of SQL Server
-                    context.setStatus("Dropping existing table");
-                    qr.update(targetConnection, "IF OBJECT_ID('dbo." + targetTable + "') IS NOT NULL DROP TABLE dbo." + targetTable);
-                }
+                if (targetSchema != null) {
```
Comment (mseaton) [inline src/main/java/org/pih/petl/job/type/SqlServerImportJob.java:116]: Per above, don't drop or create the schema if no schema has been configured.

File: README.md [hunk 14]
```diff
@@ -290,21 +300,70 @@ schedule:
...
 type: "job-pipeline"
 configuration:
   jobs:
-    - "load-upper-neno.yml"
-    - "load-lower-neno.yml"
+    - path: "load-upper-neno.yml"
+    - type: "job-pipeline"
+      configuration:
+        jobs:
+          - "load-lower-neno.yml"
```
Comment (mogoodrich) [inline README.md:326]: Maybe a super nit, but probably worth breaking inline and referenced jobs because this is a little confusing as-is (but is a rare use case)... on first read it could be interpreted that "path" and "type" apply to the same job.  Just thinking about others getting confused when reading the docs.

File: docs/examples/partitioning/README.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
+PETL - Partitioning Example
+==========================
```
Comment (mogoodrich) [inline docs/examples/partitioning/README.md:2]: Probably not worth redoing at this point, but might make sense to put the whole partitioning example in a separate commit.

File: docs/examples/partitioning/jobs/function-num-columns-changed.sql [hunk 16]
```diff
@@ -0,0 +1,22 @@
+
+-- Create function to allow determination if a schema has changed
+
+CREATE OR ALTER FUNCTION NUM_COLUMNS_CHANGED(@t1 VARCHAR(255), @t2 VARCHAR(255))
```
Comment (mogoodrich) [inline docs/examples/partitioning/jobs/function-num-columns-changed.sql:4]: Maybe a future ticket about doing a more detailed schema comparsion? I suspect that many times a schema change would only be the change of a datatype of a column... or potentially adding/removing columns at the same time and still ending up with the same number of columns... would it be straightforward to come up with a true/false type query based on something like this (would likely be database specific, but I think that's fine): https://www.mssqltips.com/sqlservertip/4824/easy-way-to-compare-sql-server-table-schemas/

Comment (mseaton) [reply docs/examples/partitioning/jobs/function-num-columns-changed.sql:4]: Sure.  This is just example code, it can be adapted and adjusted as implementers see fit.  To be clear, this function _does_ only look for datatype changes and name changes, or additions/deletions of columns.  It doesn't worry about other aspects of schema changes.  And I originally implemented this as a boolean function, but it ended up not being something I could use without comparing to 0, so I ended up making it a numeric return value to make this more explicit.

File: README.md (hunk 14 above)
Comment (mogoodrich) [reply README.md:326]: Not a strong preference, just a thought
